internal/querynodev2/delegator/ScalarPruner.go

```diff
@@ -203,62 +203,81 @@ func NewParseContext(keyField FieldID, dType schemapb.DataType) *ParseContext {
+	leftExpr, err := ParseExpr(exprPb.Left, parseCtx)
+	if err != nil {
+		return nil, err
+	}
+	rightExpr, err := ParseExpr(exprPb.Right, parseCtx)
+	if err != nil {
+		return nil, err
+	}
+	return NewLogicalBinaryExpr(leftExpr, rightExpr, exprPb.GetOp()), nil
 }
 
-func ParseLogicalUnaryExpr(exprPb *planpb.UnaryExpr, parseCtx *ParseContext) Expr {
+func ParseLogicalUnaryExpr(exprPb *planpb.UnaryExpr, parseCtx *ParseContext) (Expr, error) {
 	// currently we don't handle NOT expr, this part of code is left for logical integrity
-	return nil
+	return nil, nil
 }
 
-func ParseBinaryRangeExpr(exprPb *planpb.BinaryRangeExpr, parseCtx *ParseContext) Expr {
+func ParseBinaryRangeExpr(exprPb *planpb.BinaryRangeExpr, parseCtx *ParseContext) (Expr, error) {
 	if exprPb.GetColumnInfo().GetFieldId() != parseCtx.keyFieldIDToPrune {
-		return nil
+		return nil, nil
+	}
+	lower, err := storage.NewScalarFieldValueFromGenericValue(parseCtx.dataType, exprPb.GetLowerValue())
+	if err != nil {
+		return nil, err
 	}
-	lower := storage.NewScalarFieldValueFromGenericValue(parseCtx.dataType, exprPb.GetLowerValue())
-	upper := storage.NewScalarFieldValueFromGenericValue(parseCtx.dataType, exprPb.GetUpperValue())
-	return NewBinaryRangeExpr(lower, upper, exprPb.LowerInclusive, exprPb.UpperInclusive)
+	upper, err := storage.NewScalarFieldValueFromGenericValue(parseCtx.dataType, exprPb.GetUpperValue())
+	if err != nil {
+		return nil, err
+	}
+	return NewBinaryRangeExpr(lower, upper, exprPb.LowerInclusive, exprPb.UpperInclusive), nil
 }
 
-func ParseUnaryRangeExpr(exprPb *planpb.UnaryRangeExpr, parseCtx *ParseContext) Expr {
+func ParseUnaryRangeExpr(exprPb *planpb.UnaryRangeExpr, parseCtx *ParseContext) (Expr, error) {
 	if exprPb.GetColumnInfo().GetFieldId() != parseCtx.keyFieldIDToPrune {
-		return nil
+		return nil, nil
 	}
 	if exprPb.GetOp() == planpb.OpType_NotEqual {
-		return nil
+		return nil, nil
 		// segment-prune based on min-max cannot support not equal semantic
 	}
-	innerVal := storage.NewScalarFieldValueFromGenericValue(parseCtx.dataType, exprPb.GetValue())
-	return NewUnaryRangeExpr(innerVal, exprPb.GetOp())
+	innerVal, err := storage.NewScalarFieldValueFromGenericValue(parseCtx.dataType, exprPb.GetValue())
+	if err != nil {
+		return nil, err
+	}
+	return NewUnaryRangeExpr(innerVal, exprPb.GetOp()), nil
 }
 
-func ParseTermExpr(exprPb *planpb.TermExpr, parseCtx *ParseContext) Expr {
+func ParseTermExpr(exprPb *planpb.TermExpr, parseCtx *ParseContext) (Expr, error) {
 	if exprPb.GetColumnInfo().GetFieldId() != parseCtx.keyFieldIDToPrune {
-		return nil
+		return nil, nil
 	}
 	scalarVals := make([]storage.ScalarFieldValue, 0)
 	for _, val := range exprPb.GetValues() {
-		scalarVals = append(scalarVals, storage.NewScalarFieldValueFromGenericValue(parseCtx.dataType, val))
+		innerVal, err := storage.NewScalarFieldValueFromGenericValue(parseCtx.dataType, val)
+		if err == nil {
+			scalarVals = append(scalarVals, innerVal)
+		}
```

Comment:
handle term expr specially, because out-bound term values can be ignored directly